README.md
# CaniNES

CaniNES is a crossplatform NES emulator built on top of a FreePascal port of
the Mesen emulator core, using the SDL2 framework.

- Mesen is Copyright (C) 2014-2019 M. Bibaud, https://github.com/SourMesen/Mesen
- Currently active fork by various contributors: https://github.com/NovaSquirrel/Mesen
- nes_ntsc is Copyright (C) 2006-2007 Shay Green, http://www.slack.net/~ant/
- FreePascal ports by Joel Toivonen (hukka) 2019-2021, http://hukka.ncn.fi

Donate to the original author of Mesen: [![Donate](https://www.mesen.ca/images/donate.png)](https://www.mesen.ca/Donate.php)

## Features

- Accurate NES emulation as per Mesen
- Recent Files, Favourites, Cheat Selector
- Cartridge information screen with screenshots (requires manual configuration)
- Savestates, Rewind, Fast Forward, Run-ahead
- NTSC and CRT simulation filters

## More information

- [Latest changes](https://github.com/hukkax/CaniNES/blob/main/docs/changes.txt)
- [Known bugs and TODO list](https://github.com/hukkax/CaniNES/blob/main/docs/todo.txt)
- [Supported mappers](https://github.com/hukkax/CaniNES/blob/main/docs/mappers.txt)
- [Known nonworking ROMs](https://github.com/hukkax/CaniNES/blob/main/docs/missing.txt)

## Quick help

Press:
- Left Alt or mouse for menu
- Esc - alternative menu
- Backspace - go back to the previous page in alt. menu
- F1 - toggle info display
- F5 - save console state
- F7 - load console state
- 0-9 - select savestate slot
- F9 - file browser
- F11 - toggle full screen
- Tab - fast forward
- Backspace - rewind (if enabled in settings)
- Arrows and Z,X,Shift,Enter for NES controller 1,B,A,Select,Start;
  numpad for controller 2, or a connected joypad
- Key and joypad bindings are customizable via the Input menu

## Screenshots

![Main menu](https://github.com/hukkax/CaniNES/blob/main/docs/images/mainmenu.png)
![Cartridge information](https://github.com/hukkax/CaniNES/blob/main/docs/images/cartinfo.png)
![HEOH Demo with NTSC and CRT filters](https://github.com/hukkax/CaniNES/blob/main/docs/images/filters.png)
![BattleToads with filters disabled](https://github.com/hukkax/CaniNES/blob/main/docs/images/nofilter.png)


docs/changes.txt
=============================================================================
Changelog
=============================================================================
2021-10-13:
- The window can now be resized and maximized; removed the max scale setting
- An error dialog is now displayed in case of an initialization error

2021-10-08:
- Implemented mappers 96, 216

2021-10-07:
- Implemented mappers 82, 136, 143, 147

2021-09-30:
- Added a menubar to the window

2021-09-22:
- Initial release on GitHub

2021-09-21:
- Boxart/screenshot locations can now each contain multiple paths separated
  by semicolon; Zip archives are also supported
- The date/time format displayed when selecting a save slot can now be customized

2021-09-20:
- Files within archives are now extracted straight into memory without using
  temporary files
- If an archive only contains one ROM, automatically open it
- Archives are now colorized in the file browser

2021-09-17:
- Added support for browsing/loading ROMs from within Zip archives
- Added a scrollbar to menus when applicable
- Fixed a memory leak in file browser

2021-09-16:
- Added Favourites; use menu or Shift-F to (un)favourite, Shift-F2 to list
- Menus now colourize favourited ROMs and the currently loaded ROM
- Fixed menu rendering issues with overscan

2021-09-15:
- Toggling CRT/NTSC filters while paused/in menu now displays proper frame
- Fixed incorrect overscan bounds calculations

2021-09-14:
- All GUI colors can now be changed via the configuration file

2021-09-13:
- Improved OSD for savestate slot selection: shows availability and date
- Added OSD opacity and timeout settings
- Added setting to only show pad inputs during movie playback
- Implemented mappers 115, 193, 521

2021-09-10:
- Added setting to autoswitch 50/60 Hz display mode in fullscreen
  (switches to closest match to current desktop resolution)
- Added overscan settings
- Added alignment settings for various On-Screen Display elements
- Implemented mappers 38, 73, 78, 155, 180, 190, 250, 262, 268
- A CaniNES logo is displayed while the window is (re)initializing

2021-09-09:
- Fixed Dendy erroneously running at 60 Hz

2021-09-08:
- Optimized overlay rendering
- Refactored buggy window initialization/renderer manager code

2021-09-07:
- Added option to visualize gamepad inputs
- Refactored font rendering code

2021-09-03:
- Fixed rendering of alpha channel when using freetype

2021-09-02:
- Menus can now be operated via the mouse, right click to toggle menu

2021-09-01:
- Implemented mappers 33, 36, 37, 48, 65, 67, 93, 112, 156, 177, 184,
  87/101, 80/207, 137/138/139/141,
  157 (Datach, barcode reader not implemented),
  168 (Racermate, exercise bike not implemented)

2021-08-31:
- Implemented mappers 30, 111 (flash storage not implemented)

2021-08-30:
- Implemented mappers 70, 150, 152, 243, 185, 290
- Preliminary support for Mesen movies (MMO)
- Sped up menu generation via caching (improves ROM loading speed)

2021-08-27:
- You can now type the beginning of a title in a file listing to jump to it
- Fixed "sticking" with some shortcuts using the Shift key
- Fixed NTSC filter color burst flickering during Rewind/Fast forward
- Preliminary internal support for FCEU movies (fm2, fm3)

2021-08-25:
- Added JPEG support for Boxart/Screenshots

2021-08-24:
- Implemented Run-ahead to minimize input lag
- Implemented Freetype font rendering (optionally disabled via compiler switch)
- Added an option to change the scaling of Boxart/Screenshots
- Added an option to set the default ROM path to current dir in file browser
- Added an option to set the Boxart/Screenshot path to current dir in file
  browser (when current directory contains some png images)

2021-08-23:
- Added option to restore previous state upon loading ROM
- Added option to restore previously loaded ROM on startup
- Added option to pause emulation when in menu
- Added options related to power-on state
- Improved heuristics for finding boxart/screenshot images
- Menu rendering improvements
- Zapper is now enabled/disabled automatically, unless forced on via settings
- Fixed broken Zapper emulation

2021-08-21:
- Added support for Pro Action Rocky cheats

2021-08-20:
- Added settings for Fast forward maximum speed and frame display interval
- Added more information and gameplay screenshot to Cartridge info screen
- Improved and optimized heuristics for finding boxart/screenshot images

2021-08-19:
- Implemented mappers 172, 173
- Preliminary implementation of box art support and a Cartridge info screen
  (images must be PNG, specify directory in [Application]/ImagePath.Boxart)

2021-08-16:
- Filenames containing special characters that exist in the Latin-9 character
  set are now displayed correctly
- Menu font will now be selected and scaled according to window size
- Implemented recording audio to WAV (saved to config folder by default)

2021-08-15:
- Added Recent file history
- Added setting to use the original ROM CRC after auto-patching; enables the
  use of cheats meant for the original unpatched ROM

2021-08-14:
- Added Aspect ratio setting
- Added Filtering mode setting (nearest/linear)
- Changes to display settings are now applied on the fly
- Corresponding IPS patch is now automatically applied upon loading a ROM
  if found in the same directory
- Made the menu more aesthetically pleasing

2021-08-13:
- Implemented mapper 85 (VRC7) including OPLL sound emulation
- Implemented Jaleco mappers 18, 72, 86, 92 (no sound hardware emulation)
- Improved audio interference CPU usage

2021-08-12:
- Implemented audio interference (https://github.com/NovaSquirrel/Mesen/pull/42)
- Sped up the audio mixer

2021-08-11:
- Improved menu behavior
- Fixed Cheat code listing not clearing when changing to a ROM without cheats
- Reworked Fast forwarding to be much smoother
- Optimization to overlay drawing (don't regenerate texture if not changed)
- Added more information to the Info box display (mirroring, mapper, system)

src/core/nes.types.pas
unit NES.Types;

{$MODE DELPHI}

{$I basement.inc}

interface

uses
	Classes, SysUtils;

const
	APPNAME    = 'CaniNES';
	APPVERSION = '0.9.1';

	NES_RESOLUTION_X = 256;
	NES_RESOLUTION_Y = 240;

	CONFIG_FILENAME = APPNAME + '.ini';

	KB   = 1024;
	KB8  = KB * 8;
	KB16 = KB * 16;

	Color_Background = $FF443333;

	ArchiveFileExts = '*.zip';
	ImageFileExts   = '*.png;*.jpg;*.jpeg';
	MovieFileExts   = '*.mmo;*.fm2;*.fm3'; // .bk2
	RomFileExts     = '*.nes;*.fds' + ';' + ArchiveFileExts;
	FileBrowserExts = RomFileExts + ';' + MovieFileExts;

	ICON_SIZE    = 16;
	ICON_NONE    = 0;
	ICON_PAUSE   = 1;
	ICON_FORWARD = 2;
	ICON_REWIND  = 3;

	{$IFDEF RENDER_FONT_FREETYPE}
    TTF_FONTLICENSE = 'license.txt';

	STR_OSD_REWIND      = '◀◀';
	STR_OSD_FASTFORWARD = '▶▶';
	STR_OSD_PAUSED      = '||';
	STR_CHECKBOX_YES    = '☑ ';
	STR_CHECKBOX_NO     = '☐ ';
	STR_PAGE_NEXT       = '▶';
	STR_PAGE_PREV       = '◀ Back';
	STR_PAGE_ACCEPT     = 'OK';
	STR_ARROW_UP        = '▲';
	STR_ARROW_DOWN      = '▼';
	{$ELSE}
	STR_OSD_REWIND      = #22#23;
	STR_OSD_FASTFORWARD = #24#25;
	STR_OSD_PAUSED      = #26#27;
	STR_CHECKBOX_YES    = #20#21;
	STR_CHECKBOX_NO     = #18#19;
	STR_PAGE_NEXT       = #16;
	STR_PAGE_PREV       = #17' Back';
	{$ENDIF}

	channel_Square1      = 0;
	channel_Square2      = 1;
	channel_Triangle     = 2;
	channel_Noise        = 3;
	channel_DMC          = 4;
	channel_FDS          = 5;
	channel_VRC6         = 6;
	channel_VRC7         = 7;
	channel_MMC5         = 8;
	channel_Namco163     = 9;
	channel_Sunsoft5B    = 10;
	channel_InvA13       = 11;
	channel_InvOE1       = 12;
	channel_Last    = channel_InvOE1;
	MaxChannelCount = channel_Last;

	caBoxArt = 0;
	caSnaps  = 1;
	caTitles = 2;
	caLastIndex = caSnaps;

	irAllZeros = 0;
	irAllOnes  = 1;
	irRandom   = 2;

	INITRAM_DEFAULT = -1;
	INITRAM_RANDOM  = -2;

type
//	TByteArray = packed array of Byte;
	T2DByteArray = array of array of Byte;

	// =========
	// Console

	TNESModel   = ( nesAuto, nesNTSC, nesPAL, nesDendy );

	TGameSystem = ( NES_NTSC, NES_PAL, Famicom, Dendy, VsSystem, Playchoice, FDS, Unknown );

	// =========
	// Input

	TGameInputType = (
		gitUnspecified,
		gitStandardControllers,
		gitFourScore, gitFourPlayerAdapter,
		gitVsSystem, gitVsSystemSwapped, gitVsSystemSwapAB, gitVsZapper,
		gitZapper, gitTwoZappers,
		gitBandaiHypershot,
		gitPowerPadSideA, gitPowerPadSideB,
		gitFamilyTrainerSideA, gitFamilyTrainerSideB,
		gitArkanoidControllerNes, gitArkanoidControllerFamicom, gitDoubleArkanoidController,
		gitKonamiHyperShot,
		gitPachinkoController,
		gitExcitingBoxing,
		gitJissenMahjong,
		gitPartyTap,
		gitOekaKidsTablet,
		gitBarcodeBattler,
		gitMiraclePiano,
		gitPokkunMoguraa,
		gitTopRider,
		gitDoubleFisted,
		gitFamicom3dSystem,
		gitDoremikkoKeyboard,
		gitROB,
		gitFamicomDataRecorder,
		gitTurboFile,
		gitBattleBox,
		gitFamilyBasicKeyboard,
		gitPec586Keyboard, gitBit79Keyboard,
		gitSuborKeyboard, gitSuborKeyboardMouse1, gitSuborKeyboardMouse2,
		gitSnesMouse,
		gitGenericMulticart,
		gitSnesControllers,
		gitRacermateBicycle,
		gitUForce
	);

	// =========
	// PPU

	TPPUModel = (
		Ppu2C02,  Ppu2C03,
		Ppu2C04A, Ppu2C04B, Ppu2C04C, Ppu2C04D,
		Ppu2C05A, Ppu2C05B, Ppu2C05C, Ppu2C05D, Ppu2C05E
	);

	// =========
	// Cartridge

	TPrgMemoryType = ( PrgRom, PrgSaveRam, PrgWorkRam );
	TChrMemoryType = ( ChrDefault, ChrRom, ChrRam, ChrNametableRam );

	TBusConflictType = ( buscDefault, buscYes, buscNo );

	TMirroringType = (
		MIRROR_HORIZONTAL,
		MIRROR_VERTICAL,
		MIRROR_SCREENAONLY,
		MIRROR_SCREENBONLY,
		MIRROR_FOURSCREENS );

	// ======
	// Memory

	TMemoryOperation = ( moRead = 1, moWrite = 2, moAny = 3 );

	TMemoryOperationType = (
		memopRead,
		memopWrite,
		memopExecOpcode,
		memopExecOperand,
		memopPPURenderingRead,
		memopDummyRead,
		memopDMCRead,
		memopDummyWrite
	);

	TMemoryAccessType = (
		maUnspecified = -1,
		maNoAccess    =  0,
		maRead        =  1,
		maWrite       =  2,
		maReadWrite   =  3
	);

	TBatterySaveType = ( battNormal, battChr, battIPS );

	TAspectRatio = (
		arNoStretching,
		arAuto,
		arNTSC,
		arPAL,
		arStandard,
		arWidescreen
	);

var
	// APU channels to enable by default
	APUChannelUsed: array[0..MaxChannelCount] of Boolean = (
		True, True, True,  True, True,
		False, False, False, False, False, False,
		True, True );

const
	// controller types that use the mouse for emulated input
	MouseControllers: set of TGameInputType = [
		gitZapper, gitTwoZappers,
		gitOekaKidsTablet,
		gitSuborKeyboardMouse1, gitSuborKeyboardMouse2,
		gitSnesMouse
	];

	NESSystemNames: array[TGameSystem] of String = (
		'NES NTSC', 'NES PAL', 'Famicom', 'Dendy',
		'VsSystem', 'Playchoice', 'Famicom Disk System', 'Unknown' );

	ShortNESSystemNames: array[TGameSystem] of String = (
		'NES NTSC',
		'NES PAL',
		'Famicom',
		'Dendy',
		'Vs',
		'Playchoice',
		'FDS',
		'Unknown' );

	NESModelNames: array[TNESModel] of String = ( 'Auto', 'NTSC', 'PAL', 'Dendy' );

	PPUModelNames: array[TPPUModel] of String = (
		'Ppu2C02',  'Ppu2C03',
		'Ppu2C04A', 'Ppu2C04B', 'Ppu2C04C', 'Ppu2C04D',
		'Ppu2C05A', 'Ppu2C05B', 'Ppu2C05C', 'Ppu2C05D', 'Ppu2C05E' );

	APUChannelNames: array[0..MaxChannelCount] of String = (
	'Square 1', 'Square 2', 'Triangle', 'Noise', 'DMC',
	'FDS', 'VRC6', 'VRC7', 'MMC5', 'Namco163', 'Sunsoft5B',
	'Interference /A13', 'Interference /OE1' );

	APUChannelMixMultiplier: array[0..MaxChannelCount] of Word = (
		0, 0, 0, 0, 0, 20, 75, 43, 43, 20, 15, 26, 400 );

	MirroringTypeNames: array[TMirroringType] of String = (
		'Horizontal', 'Vertical', 'Screen A only', 'Screen B only', '4-screen' );

	ShortMirroringTypeNames: array[TMirroringType] of String = ( 'H', 'V', 'A', 'B', '4' );

	BatterySaveExtensions: array[TBatterySaveType] of String = (
		'.sav', '.chr', '.ips' );

	AspectRatioNames: array[TAspectRatio] of String = (
		'No stretching', 'Auto', 'NTSC', 'PAL', 'Standard (4:3)', 'Widescreen (16:9)' );

	function ChoosePrgMemoryType(B: Boolean; IfTrue, IfFalse: TPrgMemoryType): TPrgMemoryType; inline;
	function ChooseChrMemoryType(B: Boolean; IfTrue, IfFalse: TChrMemoryType): TChrMemoryType; inline;
	function ChooseMemoryAccessType(B: Boolean; IfTrue, IfFalse: TMemoryAccessType): TMemoryAccessType; inline;
	function ChooseMirroringType(B: Boolean; IfTrue, IfFalse: TMirroringType): TMirroringType; overload; inline;
	function ChooseMirroringType(B: Integer; IfTrue, IfFalse: TMirroringType): TMirroringType; overload; inline;
	function FileExtensionMatches(const Filename, SearchFrom: String): Boolean; inline;


implementation

function FileExtensionMatches(const Filename, SearchFrom: String): Boolean;
begin
	Result := SearchFrom.Contains(LowerCase(ExtractFileExt(Filename)));
end;

function ChoosePrgMemoryType(B: Boolean; IfTrue, IfFalse: TPrgMemoryType): TPrgMemoryType;
begin
	if B then Result := IfTrue else Result := IfFalse;
end;

function ChooseChrMemoryType(B: Boolean; IfTrue, IfFalse: TChrMemoryType): TChrMemoryType;
begin
	if B then Result := IfTrue else Result := IfFalse;
end;

function ChooseMemoryAccessType(B: Boolean; IfTrue, IfFalse: TMemoryAccessType): TMemoryAccessType;
begin
	if B then Result := IfTrue else Result := IfFalse;
end;

function ChooseMirroringType(B: Boolean; IfTrue, IfFalse: TMirroringType): TMirroringType;
begin
	if B then Result := IfTrue else Result := IfFalse;
end;

function ChooseMirroringType(B: Integer; IfTrue, IfFalse: TMirroringType): TMirroringType;
begin
	if B <> 0 then Result := IfTrue else Result := IfFalse;
end;

end.

